### PR TITLE
ztimer/core: ensure half-period tick when ztimer_now64 is used

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

ztimer has an (so far undocumented and unused) option to keep every clock's current time in 64bit.
The overflows are handled by the regular checkpointing of ztimer_now().

For 32bit clocks (clocks that are not extended), if nothing calls ztimer_now() within a period, that overflow is lost.

This PR adds a periodig half-period tick to those timers, ensuring no overflow is lost.

### Testing procedure

Tough (time consuming).

- use an empty application (where nothing calls ztimer_now()
- compile with USEMODULE+=ztimer_now64
- posssbly print the time once
- wait 2**32 us (or whatever the timebase is)
- use other means to trigger printing the current time (as 64bit value)

Without this PR, the second print should have lost the 32bit overflow.
With this PR, the time should be correctly larger than 2**32.

### CODE REVIEW

File: sys/ztimer/core.c [hunk 1]
```diff
@@ -168,6 +169,7 @@ ztimer_now_t _ztimer_now_extend(ztimer_clock_t *clock)
     clock->lower_last = lower_now;
     DEBUG("ztimer_now() returning %" PRIu32 "\n", (uint32_t)clock->checkpoint);
     ztimer_now_t now = clock->checkpoint;
+
```
Comment (bergzand) [inline sys/ztimer/core.c:172]: Can you at least put these unrelated whitespace additions in a separate commit? :)

Comment (kaspar030) [reply sys/ztimer/core.c]: sorry, amended removal of those extra lines.

**Comment by jue89** [comment]
Why don't you implement `_ztimer_update()` this way?

To me this is more readable. Furthermore, I would interpret MODULE_ZTIMER_EXTEND to extend a ztimer clock. No matter whether its a `ztimer_now64` or a `ztimer_periph_*` that doesn't support 32-bit wide timer values.

```c
static void _ztimer_update(ztimer_clock_t *clock)
{
#ifdef MODULE_ZTIMER_EXTEND
    if (clock->max_value < UINT32_MAX || IS_USED(ZTIMER_NOW64)) {
        if (clock->list.next) {
            clock->ops->set(clock,
                            _min_u32(clock->list.next->offset,
                                     clock->max_value >> 1));
        }
        else {
            clock->ops->set(clock, clock->max_value >> 1);
        }
#else
    if (0) {
#endif
    }
    else {
        clock->ops->cancel(clock);
    }
}
```

**Comment by kaspar030** [comment]
> Why don't you implement `_ztimer_update()` this way?

We could. Now ztimer_now64 depends on ztimer_extend, and _min_u32() us called in every case, whereas before it could've been skipped for non extended timers. Not sure that matters in practice, or that bit of readability has a cost.

**Review by fjmolinas — COMMENTED** [review]
I got this error:

```
/home/francisco/workspace/RIOT2/sys/ztimer/util.c:58:6: error: conflicting types for 'ztimer_periodic_wakeup'
 void ztimer_periodic_wakeup(ztimer_clock_t *clock, ztimer_now_t *last_wakeup,
      ^~~~~~~~~~~~~~~~~~~~~~
In file included from /home/francisco/workspace/RIOT2/sys/ztimer/util.c:29:
/home/francisco/workspace/RIOT2/sys/include/ztimer.h:436:6: note: previous declaration of 'ztimer_periodic_wakeup' was here
 void ztimer_periodic_wakeup(ztimer_clock_t *clock, uint32_t *last_wakeup,
```
Needs to change to `ztimer_now_t` in `ztimer.h`

**Comment by fjmolinas** [comment]
> > Why don't you implement `_ztimer_update()` this way?
> 
> We could. Now ztimer_now64 depends on ztimer_extend, and _min_u32() us called in every case, whereas before it could've been skipped for non extended timers. Not sure that matters in practice, or that bit of readability has a cost.

Isn't it still skipped if the timer is not extended? I think I'm missing somethig...

**Comment by kaspar030** [comment]
> Isn't it still skipped if the timer is not extended? I think I'm missing somethig...

no, if the timer is not extended (max_value >= UINT_MAX), it takes the second if branch, where it adds the tick if no timer is set.
if it is extended, there's already a tick.

**Comment by kaspar030** [comment]
> Needs to change to `ztimer_now_t` in `ztimer.h`

This needed a rebase, did so.

**Comment by fjmolinas** [comment]
> > Isn't it still skipped if the timer is not extended? I think I'm missing somethig...
> 
> no, if the timer is not extended (max_value >= UINT_MAX), it takes the second if branch, where it adds the tick if no timer is set.
> if it is extended, there's already a tick.

Ok, I don't have a preference regarding @jue89 suggestion, maybe @bergzand can chime in?

File: sys/ztimer/core.c [hunk 2]
```diff
@@ -296,7 +296,13 @@ static void _ztimer_update(ztimer_clock_t *clock)
             clock->ops->set(clock, clock->list.next->offset);
         }
         else {
-            clock->ops->cancel(clock);
+            if (IS_USED(ZTIMER_NOW64)) {
```
Comment (fjmolinas) [inline sys/ztimer/core.c:299]:
```suggestion
            if (IS_USED(MODULE_ZTIMER_NOW64)) {
```

Comment (kaspar030) [reply sys/ztimer/core.c]: :) nice one. fixed

**Review by fjmolinas — COMMENTED** [review]
tested and the fix works, two nodes started at the same time:

- control node

```
2020-09-23 09:38:34,201 # main(): This is RIOT! (Version: 2020.10-devel-1390-gc4b30-pr-15003)
2020-09-23 09:38:34,202 # Hello World!
2020-09-23 09:38:34,206 # You are running RIOT on a(n) p-nucleo-wb55 board.
2020-09-23 09:38:34,209 # This board features a(n) stm32 MCU.
now
2020-09-23 12:14:11,755 #  now
2020-09-23 12:14:11,757 # now: 747805186
```

- node with fix

```
2020-09-23 12:14:13,904 #  now
2020-09-23 12:14:13,906 # now: 1927344512
```

**Comment by fjmolinas** [comment]
> > Why don't you implement `_ztimer_update()` this way?
> 
> We could. Now ztimer_now64 depends on ztimer_extend, and _min_u32() us called in every case, whereas before it could've been skipped for non extended timers. Not sure that matters in practice, or that bit of readability has a cost.

So what to we do here?

**Comment by kaspar030** [comment]
> So what to we do here?

The concrete proposal might increase readability, but breaks semantics (I think). Current code is tested and works. So I'd say if someone cares, let's do a followup.

**Comment by fjmolinas** [comment]
Please squash @kaspar030

**Comment by kaspar030** [comment]
> Please squash @kaspar030

squashed

**Review by fjmolinas — APPROVED** [review]
ACK

**Comment by kaspar030** [comment]
Thanks for testing this!
